### ISSUE TITLE
Preprocessing LINEMOD Dataset Query

### ISSUE THREAD

**Issue by sangramioi** [Microsoft/singleshotpose#24]
Hello Authors,

Thank you so much for making the code available in public domain. It is very well written and intuitive.

I am sure you might have got the data from http://far.in.tum.de/Main/StefanHinterstoisser
Now, how did you:

1. Create the "masks" of the original data. 

2. Come up with labels(8 coordinates and centroid) of object in the images since they are not always at the center of the image. This could have been fairly easy if you could have used blender on mesh file to create synthetic data but how did you achieve the same thing on original images?

Please do help understand these concepts.

Thanks.

**Comment by btekin** [comment]
Thanks for your feedback on our code!

1. The masks are obtained by projecting the 3D object model onto the image with the available 6D pose of the object as in [Rad & Lepetit, ICCV'17]. Masks are only used for changing the background of the object so that the networks don't overfit to the background.

2. The labels are obtained by projecting the 3D bounding box onto the image with the ground-truth 6D pose of the object. 3D bounding box is basically the corner points of the 3D object model:  [[min_x, min_y, min_z], [min_x, min_y, max_z], [min_x, max_y, min_z], [min_x, max_y, max_z], [max_x, min_y, min_z], [max_x, min_y, max_z], [max_x, max_y, min_z], [max_x, max_y, max_z]].

**Comment by sangramioi** [comment]
Thanks for your quick response :)

Just for confirmation, the fwd projection in part 2 of my question is clear but the position of those fwd projected points in 2D images is after we have the mask, right? Since, the exact coordinates of the center object in the dataset are not given and the **images are not centered on the object.** 

_In short, in 2_. The labels are forward projection of 3D bounding box but the way we obtain their local coordinates on 2D image(not w.r.t to each other) since the images are not object centered is still confusing. 

Say this is my mesh : 

![snapshot_ape00](https://user-images.githubusercontent.com/4083183/43333791-b942a80a-91cb-11e8-9aa6-866b321bcbe7.png)

This is my fwd projection of 3D coordinates into two 2D(_I have the 8 2D coordinates too_):

![0001](https://user-images.githubusercontent.com/4083183/43333869-f8301070-91cb-11e8-9fdd-76e1b916e2dc.png)

**But how does this help on my training image label(8 coordinates in this following image) which is this, where I am not aware which part of the image is the ape object actually in**

![000064](https://user-images.githubusercontent.com/4083183/43334037-7c2bd8e6-91cc-11e8-9a74-bd1208235426.jpg)

**Comment by sangramioi** [comment]
Hey @btekin ,

Apologies for missing out on your quick response on the Issue, thanks for the help :) I had some further queries can you please help.
---------------------------------------------------------------
Just for confirmation, the fwd projection in part 2 of my question is clear but the position of those fwd projected points in 2D images is after we have the mask, right? Since, the exact coordinates of the center object in the dataset are not given and the **images are not centered on the object.** 

_In short, in 2_. The labels are forward projection of 3D bounding box but the way we obtain their local coordinates on 2D image(not w.r.t to each other) since the images are not object centered is still confusing. 

Say this is my mesh : 

![snapshot_ape00](https://user-images.githubusercontent.com/4083183/43333791-b942a80a-91cb-11e8-9aa6-866b321bcbe7.png)

This is my fwd projection of 3D coordinates into two 2D(_I have the 8 2D coordinates too_):

![0001](https://user-images.githubusercontent.com/4083183/43333869-f8301070-91cb-11e8-9fdd-76e1b916e2dc.png)

**But how does this help on my training image label(8 coordinates in this following image) which is this, where I am not aware which part of the image is the ape object actually in**

![000064](https://user-images.githubusercontent.com/4083183/43334037-7c2bd8e6-91cc-11e8-9a74-bd1208235426.jpg)

**Comment by SergioRAgostinho** [comment]
>2.  The labels are obtained by projecting the 3D bounding box onto the image with the ground-truth 6D pose of the object. 3D bounding box is basically the corner points of the 3D object model: [[min_x, min_y, min_z], [min_x, min_y, max_z], [min_x, max_y, min_z], [min_x, max_y, max_z], [max_x, min_y, min_z], [max_x, min_y, max_z], [max_x, max_y, min_z], [max_x, max_y, max_z]].

Is this the corner order you use to generate the labels? Recalling README.md 
> Respectively, 21 numbers correspond to the following: 1st number: class label, 2nd number: x0 (x-coordinate of the centroid), 3rd number: y0 (y-coordinate of the centroid), 4th number: x1 (x-coordinate of the first corner), 5th number: y1 (y-coordinate of the first corner), ..., 18th number: x8 (x-coordinate of the eighth corner), 19th number: y8 (y-coordinate of the eighth corner), 20th number: x range, 21st number: y range.

Corner 1 to 8 respectively. It might be worth to add this info to the readme.

**Comment by nkalavak** [comment]
Hi Segio and Bugra. Can either of you tell me how the labels have been assigned for the GT and prediction? For calculating R and T, I see PnP algorithm is used but I am not clear on how the order for the projections is established.

**Comment by SergioRAgostinho** [comment]
It's like Bugra wrote in his [comment](https://github.com/Microsoft/singleshotpose/issues/24#issuecomment-406578824). The bounding box corner order is this one

> [min_x, min_y, min_z], //1st corner, 4th and 5th label number
> [min_x, min_y, max_z], 
> [min_x, max_y, min_z], 
> [min_x, max_y, max_z], 
> [max_x, min_y, min_z], 
> [max_x, min_y, max_z], 
> [max_x, max_y, min_z], 
> [max_x, max_y, max_z] //8th corner, 18th and 19th label number
